File: app/src/main/java/com/memoizlabs/kotlin_playground/currency/CurrencyMapper.java
package com.memoizlabs.kotlin_playground.currency;

import rx.Observable;
import rx.functions.Func1;

public class CurrencyMapper implements GBPExchangeRateService {

    private final ExchangeRetrofitService mExchangeRetrofitService;

    public CurrencyMapper(ExchangeRetrofitService exchangeRetrofitService) {
        mExchangeRetrofitService = exchangeRetrofitService;
    }

    @Override
    public Observable<GBPExchangeRate> getRate() {
        return mExchangeRetrofitService.getRate().map(new Func1<CurrencyGson, GBPExchangeRate>() {
            @Override
            public GBPExchangeRate call(CurrencyGson currencyGson) {
                return currencyGson.toGbpExchangeRate();
            }
        });
    }
}

File: app/src/main/java/com/memoizlabs/kotlin_playground/currency/CurrencyMapper.kt
package com.memoizlabs.kotlin_playground.currency

import rx.Observable

class CurrencyMapper(private val mExchangeRetrofitService: ExchangeRetrofitService) : GBPExchangeRateService {

    override fun getRate(): Observable<GBPExchangeRate> {
        return mExchangeRetrofitService.rate().map { currencyGson -> currencyGson.toGbpExchangeRate() }
    }
}
